File: androidApp/src/main/java/org/blackcandy/android/bridge/SongsComponent.kt
package org.blackcandy.android.bridge

import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import org.blackcandy.android.fragments.web.WebFragment
import org.blackcandy.shared.viewmodels.WebViewModel

class SongsComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>,
) : BridgeComponent<HotwireDestination>(name, delegate) {
    private val viewModel: WebViewModel
        get() {
            val fragment = delegate.destination.fragment as WebFragment
            return fragment.viewModel
        }

    override fun onReceive(message: Message) {
        when (message.event) {
            "playNow" -> handlePlayNowEvent(message)
            "playNext" -> handlePlayNextEvent(message)
            "playLast" -> handlePlayLastEvent(message)
        }
    }

    private fun handlePlayNowEvent(message: Message) {
        val data = message.data<SongsData>() ?: return
        viewModel.playNow(data.songId)
    }

    private fun handlePlayNextEvent(message: Message) {
        val data = message.data<SongsData>() ?: return
        viewModel.playNext(data.songId)
    }

    private fun handlePlayLastEvent(message: Message) {
        val data = message.data<SongsData>() ?: return
        viewModel.playLast(data.songId)
    }

    @Serializable
    data class SongsData(
        val songId: Long,
    )
}

File: iosApp/iosApp/Bridge/SongsComponent.swift
import HotwireNative
import sharedKit

class SongsComponent: BridgeComponent {
    override class var name: String { "songs" }

    private var viewModel: WebViewModel? {
        let viewController = delegate?.destination as? WebViewController
        return viewController?.viewModel
    }

    override func onReceive(message: Message) {
        switch message.event {
        case "playNow":
            handlePlayNowEvent(message)
        case "playNext":
            handlePlayNextEvent(message)
        case "playLast":
            handlePlayLastEvent(message)
        default:
            break
        }
    }

    private func handlePlayNowEvent(_ message: Message) {
        guard let data: SongsData = message.data() else { return }
        viewModel?.playNow(songId: data.songId)
    }

    private func handlePlayNextEvent(_ message: Message) {
        guard let data: SongsData = message.data() else { return }
        viewModel?.playNext(songId: data.songId)
    }

    private func handlePlayLastEvent(_ message: Message) {
        guard let data: SongsData = message.data() else { return }
        viewModel?.playLast(songId: data.songId)
    }
}

extension SongsComponent {
    struct SongsData: Decodable {
        let songId: Int64
    }
}
